gcs/indexer.py

```diff
@@ -37,37 +38,58 @@ def parse_args():
     return parser.parse_args()
 
 
-def index_gcs_pattern(gcs_pattern):
-    # Input gcs_pattern looks like
-    # gs://genomics-public-data/1000-genomes/bam/PRIMARY_KEY.
-
+def index_gcs_files(es, index_name, gcs_pattern):
     logger.info('Processing %s' % gcs_pattern)
 
+    # Input gcs_pattern looks like
+    # gs://genomics-public-data/1000-genomes/bam/PRIMARY_KEY.
     trimmed_gcs_pattern = gcs_pattern.replace('gs://', '')
-
     # bucket_str looks like genomics-public-data
     bucket_str = trimmed_gcs_pattern.split('/')[0]
-
-    prefix = trimmed_gcs_pattern.split('/', 1)[1]
     # prefix looks like 1000-genomes/bam/
+    prefix = trimmed_gcs_pattern.split('/', 1)[1]
     prefix = prefix[:prefix.index('PRIMARY_KEY')]
 
     logger.info('Retrieving objects from bucket %s with prefix %s.' %
                 (bucket_str, prefix))
     bucket = storage.Client(project=None).bucket(bucket_str)
     objects = bucket.list_blobs(prefix=prefix)
-
     regex = re.compile(gcs_pattern.replace('PRIMARY_KEY', '(\w+)'))
 
+    # Group each object by primary key and file type, e.g.
+    # {
+    #   'PRIMARY_KEY1' : {
+    #       'bam': {
+    #           'count': 2,
+    #           'files': ['gs://b/file1.bam', 'gs://b/file2.bam'],
+    #       'vcf', {...}
+    #   }, ...
+    # }
+    file_sets = {}
     for obj in objects:
-        obj_path = 'gs://%s/%s' % (bucket_str, obj.name)
-        match = re.match(regex, obj_path)
-        if match:
-            primary_key = match.group(1)
-            print(
-                'Identified primary key %s from %s' % (primary_key, obj_path))
-        else:
-            raise ValueError('Could not find primary key in %s' % obj_path)
+        path = 'gs://%s/%s' % (bucket_str, obj.name)
+        match = re.match(regex, path)
+        if not match:
+            continue
+        key = match.group(1)
+        file_type = ''
+        for t in FILE_TYPES:
+            if t in path:
+                file_type = t
+        if not file_type:
+            continue
+        _append_to_file_sets(file_sets, key, file_type, path)
+
+    indexer_util.bulk_index(es, index_name, file_sets.iteritems())
+
+
+def _append_to_file_sets(file_sets, key, file_type, path):
```

Comment:
Nit in these repos, we use the convention that helper methods are declared before they're used. Can you move `_append_to_file_sets` to before `_index_gcs_files`?

Reply:
Done.